### PR TITLE
Audio viewer Safari problem

### PR DESCRIPTION
**Pull request by aorin**
Problem: Audio viewer didn't work at all with older Safari versions

Solution: Problem was caused by old Safari not supporting loading/resampling audio into a sample rate we are using. I noticed that it wasn't necessary to resample the audio in order to draw spectrogram (as long as `nperseg` and `noverlap` parameters take into account the sample rate). Now the audio viewer should work for all browsers (that support Web Audio API).

This pull request is bigger than it should be because I wasn't sure at first how to fix this. I also did some refactoring.

Testable here: https://180916086.dev.laji.fi/theme/kerttu/letters (should look the same than dev except if you are using an old Safari)

### CODE REVIEW

File: projects/laji/src/app/shared-modules/audio-viewer/audio-viewer/audio-viewer.component.ts [hunk 1]
```diff
@@ -49,8 +49,8 @@ export class AudioViewerComponent implements OnChanges, OnDestroy {
 
   @Input() spectrogramConfig: ISpectrogramConfig = {
     sampleRate: 22050,
-    nperseg: 256,
-    noverlap: 256 - 160,
+    targetWindowLengthInSeconds: 0.015,
+    targetWindowOverlapPercentage: 0.375,
```
Comment (olzraiti) [inline projects/laji/src/app/shared-modules/audio-viewer/audio-viewer/audio-viewer.component.ts:53]: These are already in variables file, could they be imported from there?

**Review by olzraiti — CHANGES_REQUESTED** [review]
Looks good, left some minor nits.

File: projects/laji/src/app/shared-modules/audio-viewer/audio-viewer/audio-spectrogram/audio-spectrogram.component.ts [hunk 2]
```diff
@@ -98,11 +99,13 @@ export class AudioSpectrogramComponent implements AfterViewInit, OnChanges {
...
-  private updateWidthAndHeigth() {
+  private updateWidthAndHeight() {
     this._width = this.width
       ? this.width : this.containerRef
       ? Math.max(this.containerRef.nativeElement.offsetWidth - this._margin.left - this._margin.right, 0)
       : 0;
-    this._height = this.height ? this.height : this.config ? this.config.nperseg / 2 : 0;
+    this._height = this.height ? this.height : this.config ? (
+      AudioViewerUtils.getSpectrogramSegmentLength(this.config.targetWindowLengthInSeconds, this.config.sampleRate) / 2
+    ) : 0;
```
[the remark this answers is not in the document]
Comment (olzraiti) [reply projects/laji/src/app/shared-modules/audio-viewer/audio-viewer/audio-spectrogram/audio-spectrogram.component.ts]: Now you made it like this:

```
a
  ? b : c
  ? d
  : e
```

which is inconsistent. I think having each `?` and `:` on a separate line, and indenting systematically makes it visually easy to understand, like this:

```
  this._height = this.height
    ? this.height
    : this.config
      ? (AudioViewerUtils.getSpectrogramSegmentLength(this.config.targetWindowLengthInSeconds, this.config.sampleRate) / 2)
      : 0;
```

(I didn't notice that width calculation above this has the same problem, maybe we should fix that also to make in consistent)

[the remark this answers is not in the document]
Comment (aorin) [reply projects/laji/src/app/shared-modules/audio-viewer/audio-viewer/audio-spectrogram/audio-spectrogram.component.ts]: Ok, let's change it to that. I'm pretty sure that the format in`this._width` has been copied somewhere else, trying to mimic our default format (but we probably doesn't have any default format in this yet)

**Comment by olzraiti** [comment]
Thanks!

**Review by olzraiti — APPROVED** [review]

**Pull request merged**